### ISSUE TITLE
Build failing (compilation of rsdd)

### ISSUE THREAD

**Issue by fzaiser** [SHoltzen/dice#65]
I followed the build instructions, but hitting the following error when it tries to compile rsdd:

```
[...]
   Compiling rsdd v0.1.0 (/home/fabian/sync/repo/dice/_build/default/rsdd)
warning: trait objects without an explicit `dyn` are deprecated
  --> src/sample/random.rs:75:35
   |
75 |     pub fn bind<R>(&self, f: &mut FnMut(&T) -> Random<R>) -> Random<R> {
   |                                   ^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this is accepted in the current edition (Rust 2018) but is a hard error in Rust 2021!
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
   = note: `#[warn(bare_trait_objects)]` on by default
help: use `dyn`
   |
75 |     pub fn bind<R>(&self, f: &mut dyn FnMut(&T) -> Random<R>) -> Random<R> {
   |                                   +++

warning: `rsdd` (lib) generated 1 warning (run `cargo fix --lib -p rsdd` to apply 1 suggestion)
error[E0599]: no method named `get_all_assignments` found for reference `&BayesianNetwork` in the current scope
  --> bin/bayesian_network_compiler.rs:44:37
   |
44 |             for varassgn in network.get_all_assignments(&v) {
   |                                     ^^^^^^^^^^^^^^^^^^^ method not found in `&BayesianNetwork`

error[E0599]: no method named `get_conditional_prob` found for reference `&BayesianNetwork` in the current scope
  --> bin/bayesian_network_compiler.rs:57:44
   |
57 |                     let cur_prob = network.get_conditional_prob(&v, varassgn, &passgn);
   |                                            ^^^^^^^^^^^^^^^^^^^^ method not found in `&BayesianNetwork`

error[E0599]: no method named `get_all_assignments` found for reference `&BayesianNetwork` in the current scope
  --> bin/compare_canonicalize.rs:76:33
   |
76 |         for varassgn in network.get_all_assignments(&v) {
   |                                 ^^^^^^^^^^^^^^^^^^^ method not found in `&BayesianNetwork`

error[E0599]: no function or associated item named `from_string` found for struct `BayesianNetwork` in the current scope
   --> bin/compare_canonicalize.rs:136:31
    |
136 |     let bn = BayesianNetwork::from_string(std::fs::read_to_string(args.file).unwrap().as_str());
    |                               ^^^^^^^^^^^ function or associated item not found in `BayesianNetwork`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rsdd` (bin "compare_canonicalize") due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error[E0599]: no method named `get_all_assignments` found for struct `BayesianNetwork` in the current scope
   --> bin/bayesian_network_compiler.rs:211:33
    |
211 |         for varassgn in network.get_all_assignments(&v) {
    |                                 ^^^^^^^^^^^^^^^^^^^ method not found in `BayesianNetwork`

error[E0599]: no method named `get_conditional_prob` found for struct `BayesianNetwork` in the current scope
   --> bin/bayesian_network_compiler.rs:224:40
    |
224 |                 let cur_prob = network.get_conditional_prob(&v, varassgn, &passgn);
    |                                        ^^^^^^^^^^^^^^^^^^^^ method not found in `BayesianNetwork`

error[E0599]: no function or associated item named `from_string` found for struct `BayesianNetwork` in the current scope
   --> bin/bayesian_network_compiler.rs:315:31
    |
315 |     let bn = BayesianNetwork::from_string(std::fs::read_to_string(&args.file).unwrap().as_str());
    |                               ^^^^^^^^^^^ function or associated item not found in `BayesianNetwork`

error: could not compile `rsdd` (bin "bayesian_network_compiler") due to 5 previous errors
```

My rustc version is 1.72.1 (d5c2e9c34 2023-09-13).

**Comment by SHoltzen** [comment]
Thanks for your interest in using Dice.

I was able to reproduce this error. It is due to a stale dependency in the `rsdd` repo. This is one of the things that CI does not catch unfortunately.

This was resolved in #ed86716 by creating a new `dice-rsdd` branch in the `rsdd` repo, and updating  (cc @mattxwang  do you know of a better solution here long-term for maintaining submodule compatibility between these repos? In the mean-time).

Let me know if there are further issues.

Cheers,
Steven

**Comment by mattxwang** [comment]
Thanks for giving dice a spin @fzaiser - let us know if there are still problems!

Re: 

> This was resolved in #ed86716 by creating a new `dice-rsdd` branch in the `rsdd` repo, and updating (cc @mattxwang do you know of a better solution here long-term for maintaining submodule compatibility between these repos? In the mean-time).

Now that RSDD is on [opam](https://opam.ocaml.org/packages/rsdd) (via [rsdd-ocaml](https://github.com/neuppl/rsdd-ocaml)), I think the correct path forward is to have opam install it as a package (instead of using a submodule). I can work on that!

**Comment by fzaiser** [comment]
Thanks for the quick fix! I can build and run Dice now :)
